### PR TITLE
Add diff and rollback for migrations

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
This PR adds the functionality for getting a diff for the current schema and the source. Also adds a rollback method a.k.a. `GetOppositeMigrations` to calculate required migrations to restore existing state.

Also, merge CI tests on one step, we don't actually need to run on different setups.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-47971

### CODE REVIEW

**Comment by agnivade** [comment]
I had something a bit different than this in mind. Our main objective is to just let the SRE team the ability to rollback whatever migrations were applied in the current upgrade.

This means rolling back already applied, and un-applied versions as well. For this, we need to have a separate checkpointing mechanism. Because every upgrade can contain multiple migrations, we cannot rely on migrations as the single unit of change. 

If an upgrade has 5 migrations, 2 were successful, and 1 failed, we want to roll back all 3 migrations. For that, we need to load the last successful checkpoint, and only save the current checkpoint when all 5 migrations were complete. And if there are any failures, run the reverse migrations from the currently failed migration to the last successful checkpoint.

Basically, something like:

ApplyAll()
|
- if no error -> SaveCheckpoint()
- if error -> LoadCheckpoint(), calculate diff of applied migrations till that checkpoint, apply down migrations for those.

Of course, there needs to be atleast one checkpoint present first. I think that is acceptable.

**Comment by isacikgoz** [comment]
@agnivade This is a preliminary for the rollback logic. The next PR is aiming to have a similar functionality you mentioned. In fact let me add those to this PR.

**Comment by isacikgoz** [comment]
> @isacikgoz - is this ready to review?

Yep

File: models/plan.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+		RevertMigrations: rollback,
+	}
+}
+
+func (p *Plan) Validate() error {
+	if p.Version != CurrentPlanVersion {
+		return ErrInvalidPlanVersion
+	}
+
+	return nil
```
Comment (agnivade) [inline models/plan.go:33]: What is the objective behind this? Any reason to keep a version number?

File: morph.go [hunk 2]
```diff
@@ -209,23 +217,157 @@ func (m *Morph) ApplyDown(limit int) (int, error) {
...
+	revertMigrations := make([]*models.Migration, 0, len(plan.RevertMigrations))
+
+	for i := range plan.Migrations {
+		// add to the revert queue
+		for _, migration := range plan.RevertMigrations {
+			if migration.Name == plan.Migrations[i].Name {
+				revertMigrations = append(revertMigrations, migration)
+				break
+			}
+		}
```
Comment (agnivade) [inline morph.go:336]: Any reason to get this separately again? In `NewPlan`, can't we validate that len(migrations) == len(reverseMIgrations), and the name of each migration at index `i` is the same? Then we don't need all this

File: morph.go [hunk 3]
```diff
@@ -209,23 +217,157 @@ func (m *Morph) ApplyDown(limit int) (int, error) {
...
+	pendingMigrations, err := computePendingMigrations(appliedMigrations, m.source.Migrations())
+	if err != nil {
+		return nil, err
+	}
+
+	var diff []*models.Migration
+	for _, migration := range sortMigrations(pendingMigrations) {
+		if migration.Direction != models.Up {
+			continue
+		}
```
Comment (agnivade) [inline morph.go:261]: driver.AppliedMigrations always returns Up migrations. https://github.com/mattermost/morph/blob/d876fb543796cded9a08046ff218c80832ecdf0f/drivers/postgres/postgres.go#L289. Same for MySQL as well. Is there any need for this check?

File: morph.go [hunk 4]
```diff
@@ -209,23 +217,157 @@ func (m *Morph) ApplyDown(limit int) (int, error) {
...
+			return nil, err
+		}
+
+		diff := make([]*models.Migration, 0, len(downMigrations))
+		for i := 0; i < len(sortedMigrations); i++ {
+			diff = append(diff, downMigrations[sortedMigrations[i].Name])
+		}
+
+		return diff, nil
+	}
```
Comment (agnivade) [inline morph.go:250]: I am trying to understand if we would ever need this? This just returns the down migrations for _all_ applied forward migrations. This is exactly the same code as `ApplyDown(-1)`. 

Whereas the forward diff actually calculates the diff between what's pending to be applied vs what's already applied

File: morph.go [hunk 5]
```diff
@@ -209,23 +217,157 @@ func (m *Morph) ApplyDown(limit int) (int, error) {
...
+				// So in this case, we need to apply the migration_2 (up) but it will be in the migrations table.
+				// Therefore we are not saving the version in the database because it will fail on the save version step.
+				skipSave := revertMigrations[j].Direction == models.Up && j == len(revertMigrations)-1
+				rErr := m.apply(revertMigrations[j], !skipSave)
+				if rErr != nil {
+					return fmt.Errorf("could not rollback migrations after trying to migrate: %w", rErr)
+				}
+
+				m.config.Logger.Printf("successfully rolled back migration: %s", revertMigrations[j].Name)
+			}
```
Comment (agnivade) [inline morph.go:361]: For better readability, I'd suggest to break the parent loop on error, and note the index till which it ran. Then move this loop outisde.

File: morph.go [hunk 6]
```diff
@@ -209,23 +217,157 @@ func (m *Morph) ApplyDown(limit int) (int, error) {
...
+				break
+			}
+		}
+
+		err := m.apply(plan.Migrations[i], true)
+		if err != nil {
+			// log the fail step
+			m.config.Logger.Printf("migration %s failed, starting rollback", plan.Migrations[i].Name)
+
+			for j := len(revertMigrations) - 1; j >= 0; j-- {
```
Comment (agnivade) [inline morph.go:343]: Any reason we are starting from the very end, if we know that `i` might be < `len(revertMigrations)`?

This is what I was thinking:

if we have migration 1, 2, 3. Then 1 passed, 2 failed, we will just run revert migrations from 2,1. In your case, we are running reverse migrations for 3,2,1.

File: morph.go [hunk 7]
```diff
@@ -209,23 +217,157 @@ func (m *Morph) ApplyDown(limit int) (int, error) {
...
+				// There is a special case when we are reverting a rollback
+				// We shouldn't save the version if we are trying to restore the last applied migration
+				// here is an example, lets say we have following migrations in the applied migrations table:
+				// migration_1, migration_2, migration_3
+				// Once we initiate the rollback, we will have the following:
+				// migration_3, migration_2, migration_1 (to rollback)
+				// Let's say we have a bug in migration_2 and failed.
+				// We don't remove that version from the database, because migration is not successfully rolled back.
+				// So in this case, we need to apply the migration_2 (up) but it will be in the migrations table.
+				// Therefore we are not saving the version in the database because it will fail on the save version step.
```
Comment (agnivade) [inline morph.go:353]: Can you clarify this a bit for me?

File: morph.go (hunk 2 above)
Comment (isacikgoz) [reply morph.go:336]: So the logic here is to have a stack of revert migrations. We push migrations here and then read in reverse. We are just making sure that we are setting the right order of what we have applied or attempted to apply.

I think we can lean on the order in the plan, but this is just safer to me.

File: morph.go (hunk 3 above)
Comment (isacikgoz) [reply morph.go:261]: Makes sense, I was being a bit cautious.

File: morph.go (hunk 4 above)
Comment (isacikgoz) [reply morph.go:250]: >  This is exactly the same code as ApplyDown(-1)

Not if we limit the migrations. Let's say you want to rollback 3 migrations. You get the diff (in this case all of the applied), the you pick last 3 migrations, generate the plan accordingly and apply it.

File: morph.go (hunk 6 above)
Comment (isacikgoz) [reply morph.go:343]: Since we queued the revert migrations, we are reading it from the last to treat it as a stack. In my case it will work like this:
- add revert migration_1 to the queue
- apply migration 1
- no error continue
- add revert migration_2 to the queue
- apply migration 2
- error ->
  - apply revert migration_2 (since we are reading in reverse order)
  - apply revert migration_1
- return error

File: morph.go (hunk 5 above)
Comment (isacikgoz) [reply morph.go:361]: Sure 👍

File: morph.go (hunk 7 above)
Comment (isacikgoz) [reply morph.go:353]: So in this scenario where we are actually trying rolling back. I mean the direction of the plan is `down`. But since we failed in the migration (which is a `down` script) phase we couldn't remove the applied migration from the `db_migrations` table. In that case if we try to save the version, because in the next step we are applying an `up` script, it will fail.

File: morph.go (hunk 4 above)
Comment (agnivade) [reply morph.go:250]: > You get the diff (in this case all of the applied),

But all of the applied will include all historical migrations as well right?

File: models/plan.go (hunk 1 above)
Comment (agnivade) [reply models/plan.go:33]: Do you see a use-case in serializing a plan to file and then executing it?

Comment (isacikgoz) [reply models/plan.go:33]: Yeah, actually the #21 works that way.

File: morph.go (hunk 4 above)
Comment (isacikgoz) [reply morph.go:250]: That's correct. But that's the logical diff if you look at the opposite direction :p

File: models/plan.go (hunk 1 above)
Comment (agnivade) [reply models/plan.go:33]: Gotcha, yep makes sense.

File: morph.go (hunk 4 above)
Comment (agnivade) [reply morph.go:250]: Right, so my point is - will this be helpful in real life? We are already getting the reverse migrations by creating the plan.

File: morph.go (hunk 6 above)
Comment (agnivade) [reply morph.go]: Ah okay, I read the code incorrectly. So I was thinking of just using the index till where the forward loop ran, and then just use that index to traverse the reverseMigrations slices backwards. Then you don't need to create a stack and pop them. This felt a bit simpler mentally to me. But I'll leave it up to you.

File: morph.go (hunk 7 above)
Comment (agnivade) [reply morph.go]: resolved offline

File: morph.go (hunk 4 above)
Comment (isacikgoz) [reply morph.go:250]: Yes if you plan to dowgrade. You get the diff, you limit the diff and generate the plan. The other alternative is to ask for limit in this method. But I wanted to keep this simple.

File: morph.go (hunk 2 above)
Comment (agnivade) [reply morph.go]: My mistake, see https://github.com/mattermost/morph/pull/17#discussion_r1022447007

File: morph.go (hunk 6 above)
Comment (isacikgoz) [reply morph.go]: That's indeed much simpler, but it leans on the *right* order of the plan. We are now ensuring the order is right. That's why added the logic.
